lib/workers.js

```diff
@@ -0,0 +1,294 @@
+        throw new Error('Test group should be an array');
+      }
+      for (const testGroup of testGroups) {
+        this.testGroups.push(convertToMochaTests(testGroup));
+      }
+    } else if (typeof numberOfWorkers === 'number' && numberOfWorkers > 0) {
+      this.testGroups = config.by === 'suite' ? this.createGroupsOfSuites(numberOfWorkers) : this.createGroupsOfTests(numberOfWorkers);
+    }
+
+    this.workers = createWorkerObjects(this.testGroups, this.codecept.config, config.testConfig);
+    this.numberOfWorkers = this.workers.length;
+  }
+
+  /**
+   * Creates a new worker
+   *
+   * @returns {WorkerObject}
+   */
+  spawn() {
+    const worker = new WorkerObject(this.numberOfWorkers);
+    this.workers.push(worker);
+    this.numberOfWorkers += 1;
+    return worker;
+  }
+
+  /**
+   * @param {Number} numberOfWorkers
+   */
+  createGroupsOfTests(numberOfWorkers) {
+    const files = this.codecept.testFiles;
+    const mocha = Container.mocha();
+    mocha.files = files;
+    mocha.loadFiles();
+    const groups = populateGroups(numberOfWorkers);
+    let groupCounter = 0;
+
+    mocha.suite.eachTest((test) => {
+      const i = groupCounter % groups.length;
+      if (test) {
+        const { id } = test;
+        groups[i].push(id);
+        groupCounter++;
+      }
+    });
+    return groups;
+  }
+
+  /**
+   * @param {Number} numberOfWorkers
+   */
+  createGroupsOfSuites(numberOfWorkers) {
+    const files = this.codecept.testFiles;
+    const mocha = Container.mocha();
+    mocha.files = files;
+    mocha.loadFiles();
+    const groups = populateGroups(numberOfWorkers);
+
+    mocha.suite.suites.forEach((suite) => {
+      const i = indexOfSmallestElement(groups);
+      suite.tests.forEach((test) => {
+        if (test) {
+          const { id } = test;
+          groups[i].push(id);
+        }
+      });
+    });
+    return groups;
+  }
+
+  /**
+   * @param {Object} config
+   */
+  overrideConfig(config) {
+    for (const worker of this.workers) {
+      worker.addConfig(config);
+    }
+  }
+
+  run() {
+    runHook(this.codecept.config.bootstrapAll, () => {
+      for (const worker of this.workers) {
+        const workerThread = createWorker(worker);
+        this._listenWorkerEvents(workerThread);
+      }
+    }, 'bootstrapAll');
+
+    return this;
+  }
+
+  /**
+   * @returns {Array<WorkerObject>}
+   */
+  getWorkers() {
+    return this.workers;
+  }
+
+  /**
+   * @returns {Boolean}
+   */
+  isFailed() {
+    return (this.stats.failures || this.errors.length) > 0;
+  }
+
+  _listenWorkerEvents(worker) {
+    worker.on('message', (message) => {
+      output.process(message.workerIndex);
+      switch (message.event) {
+        case event.test.failed: this.emit(event.test.failed, repackTest(message.data)); break;
+        case event.test.passed: this.emit(event.test.passed, repackTest(message.data)); break;
+        case event.all.after: this._appendStats(message.data); break;
```

Comment:
Could you also add test.finished and test.skipped (if this event exists, I am not sure) event.

Reply:
@DavertMik test.finished is available, I will add it